src/consensus/blockstore.rs

```diff
@@ -265,70 +138,105 @@ impl Blockstore {
     /// Returns `None` if blockstore does not hold any version of the block yet.
     #[must_use]
     pub fn canonical_block_hash(&self, slot: Slot) -> Option<Hash> {
-        self.canonical.get(&slot).copied()
+        self.slot_data(slot)?
+            .canonical
+            .completed
+            .clone()
+            .map(|c| c.0)
     }
 
-    /// Gives any relevant alternative block hashes for a given slot, if any.
-    #[must_use]
-    pub fn alternative_block_hashes(&self, slot: Slot) -> Option<&[Hash]> {
-        self.alternatives.get(&slot).map(|v| v.as_ref())
+    /// Gives the number of stored slices for a given `slot`.
+    pub fn stored_slices_for_slot(&self, slot: Slot) -> usize {
+        self.slot_data(slot).map_or(0, |s| s.canonical.slices.len())
+    }
+
+    /// Gives the number of stored shreds for a given `slot` (across all slices).
+    pub fn stored_shreds_for_slot(&self, slot: Slot) -> usize {
+        self.slot_data(slot).map_or(0, |s| {
+            s.canonical.shreds.values().map(Vec::len).sum::<usize>()
```

Comment:
```suggestion
            s.canonical.shreds.values().map(Vec::len).sum()
```

I think the compiler should be able to pick this up.